# Pulizia e preparazione del dataset di magazzino

In questo notebook viene eseguita la pulizia iniziale del dataset di magazzino.

L'obiettivo è correggere eventuali problemi nei dati, controllare valori mancanti e duplicati, uniformare i testi, convertire correttamente prezzi e date e preparare un dataset pulito da utilizzare nelle analisi successive.

In [1]:
#Importazione delle librerie per l'elaborazione dei dati:

import pandas as pd
import numpy as np
from pathlib import Path


In [2]:
# Verifica della cartella di lavoro e presenza del dataset originale:

file_input = Path("Dataset.csv")

print("Cartella di lavoro:")
print(Path.cwd())

print("\nPercorso del dataset:")
print(file_input.resolve())

print("\nDataset presente?")
print(file_input.exists())


Cartella di lavoro:
C:\Users\mace\OneDrive\python\installazione1\project work\Analisi dati di magazzino

Percorso del dataset:
C:\Users\mace\OneDrive\python\installazione1\project work\Analisi dati di magazzino\Dataset.csv

Dataset presente?
True


In [3]:
# Importazione del dataset:

df = pd.read_csv(file_input)

print("Dataset importato.")
print("\nPrime 10 righe del dataset:")

df.head(10)


Dataset importato.

Prime 10 righe del dataset:


,Product_ID,Product_Name,Catagory,Supplier_ID,Supplier_Name,Stock_Quantity,Reorder_Level,Reorder_Quantity,Unit_Price,Date_Received,Last_Order_Date,Expiration_Date,Warehouse_Location,Sales_Volume,Inventory_Turnover_Rate,Status
0,29-205-1132,Sushi Rice,Grains & Pulses,38-037-1699,Jaxnation,22,72,70,$4.50,8/16/2024,6/29/2024,9/19/2024,48 Del Sol Trail,32,19,Discontinued
1,40-681-9981,Arabica Coffee,Beverages,54-470-2479,Feedmix,45,77,2,$20.00,11/1/2024,5/29/2024,5/8/2024,36 3rd Place,85,1,Discontinued
2,06-955-3428,Black Rice,Grains & Pulses,54-031-2945,Vinder,30,38,83,$6.00,8/3/2024,6/10/2024,9/22/2024,3296 Walton Court,31,34,Backordered
3,71-594-6552,Long Grain Rice,Grains & Pulses,63-492-7603,Brightbean,12,59,62,$1.50,12/8/2024,2/19/2025,4/17/2024,3 Westerfield Crossing,95,99,Active
4,57-437-1828,Plum,Fruits & Vegetables,54-226-4308,Topicstorm,37,30,74,$4.00,7/3/2024,10/11/2024,10/5/2024,15068 Scoville Court,62,25,Backordered
5,21-120-6238,All-Purpose Flour,Grains & Pulses,86-292-4587,Dabjam,55,33,14,$1.75,12/3/2024,5/26/2024,9/5/2024,050 Mcbride Avenue,34,62,Discontinued
6,71-516-1996,Corn Oil,Oils & Fats,04-391-7610,Tagfeed,96,52,16,$2.50,3/18/2024,5/7/2024,6/20/2024,12 Truax Court,67,13,Active
7,39-629-5554,Egg (Goose),Dairy,67-679-4930,Muxo,44,90,17,$2.50,2/3/2025,4/9/2024,2/5/2025,267 International Plaza,21,91,Discontinued
8,66-268-8345,Greek Yogurt,Dairy,32-182-1895,Thoughtstorm,91,84,11,$3.00,12/4/2024,6/2/2024,1/8/2025,550 Clemons Plaza,56,90,Active
9,46-452-9419,Egg (Duck),Dairy,67-137-4215,Wordify,43,10,15,$1.00,11/18/2024,11/14/2024,7/8/2024,55782 Welch Hill,27,69,Active


In [4]:
# Controllo dimensioni del dataset

print("Numero di righe e colonne:")
print(df.shape)

print("\nNomi delle colonne:")
print(df.columns.tolist())


Numero di righe e colonne:
(990, 16)

Nomi delle colonne:
['Product_ID', 'Product_Name', 'Catagory', 'Supplier_ID', 'Supplier_Name', 'Stock_Quantity', 'Reorder_Level', 'Reorder_Quantity', 'Unit_Price', 'Date_Received', 'Last_Order_Date', 'Expiration_Date', 'Warehouse_Location', 'Sales_Volume', 'Inventory_Turnover_Rate', 'Status']


In [5]:
# Controllo della struttura e dei tipi di dato

df.info()

<class 'pandas.DataFrame'>
RangeIndex: 990 entries, 0 to 989
Data columns (total 16 columns):
 #   Column                   Non-Null Count  Dtype
---  ------                   --------------  -----
 0   Product_ID               990 non-null    str  
 1   Product_Name             990 non-null    str  
 2   Catagory                 989 non-null    str  
 3   Supplier_ID              990 non-null    str  
 4   Supplier_Name            990 non-null    str  
 5   Stock_Quantity           990 non-null    int64
 6   Reorder_Level            990 non-null    int64
 7   Reorder_Quantity         990 non-null    int64
 8   Unit_Price               990 non-null    str  
 9   Date_Received            990 non-null    str  
 10  Last_Order_Date          990 non-null    str  
 11  Expiration_Date          990 non-null    str  
 12  Warehouse_Location       990 non-null    str  
 13  Sales_Volume             990 non-null    int64
 14  Inventory_Turnover_Rate  990 non-null    int64
 15  Status           

---

E' stato individuato un errore di nomenclatura dell attributo categoria, erronemanete definito: "Catagory"



In [6]:
# Correzione del nome della colonna "Category"

df = df.rename(columns={"Catagory": "Category"})

print("Nomi delle colonne:")
print(df.columns.tolist())


Nomi delle colonne:
['Product_ID', 'Product_Name', 'Category', 'Supplier_ID', 'Supplier_Name', 'Stock_Quantity', 'Reorder_Level', 'Reorder_Quantity', 'Unit_Price', 'Date_Received', 'Last_Order_Date', 'Expiration_Date', 'Warehouse_Location', 'Sales_Volume', 'Inventory_Turnover_Rate', 'Status']


In [7]:
# Controllo dei valori mancanti per ogni colonna

valori_mancanti = df.isna().sum()

print("Valori mancanti per colonna:")
print(valori_mancanti)


Valori mancanti per colonna:
Product_ID                 0
Product_Name               0
Category                   1
Supplier_ID                0
Supplier_Name              0
Stock_Quantity             0
Reorder_Level              0
Reorder_Quantity           0
Unit_Price                 0
Date_Received              0
Last_Order_Date            0
Expiration_Date            0
Warehouse_Location         0
Sales_Volume               0
Inventory_Turnover_Rate    0
Status                     0
dtype: int64



---

E' stato identificato un potenziale valore mancante nella colonna delle categorie. Di seguito verrà analizzato nel dettaglio questo elemento


In [8]:
# Visualizzazione delle righe che contengono almeno un valore mancante

righe_con_valori_mancanti = df[df.isna().any(axis=1)]

print("Righe con almeno un valore mancante:")
righe_con_valori_mancanti


Righe con almeno un valore mancante:


,Product_ID,Product_Name,Category,Supplier_ID,Supplier_Name,Stock_Quantity,Reorder_Level,Reorder_Quantity,Unit_Price,Date_Received,Last_Order_Date,Expiration_Date,Warehouse_Location,Sales_Volume,Inventory_Turnover_Rate,Status
685,10-378-9729,Cabbage,NaN,83-941-9620,Rooxo,69,21,68,$66.55,12/23/2024,11/26/2024,9/21/2024,2 Butterfield Pass,36,35,Discontinued


In [9]:
# Elenco delle categorie presenti nel dataset

print("Categorie presenti:")
print(df["Category"].value_counts(dropna=False))


Categorie presenti:
Category
Fruits & Vegetables    331
Dairy                  180
Grains & Pulses        162
Seafood                 90
Oils & Fats             77
Beverages               75
Bakery                  74
NaN                      1
Name: count, dtype: int64


In [10]:
# Assegnazione della categoria corretta al prodotto Cabbage

# df.loc[righe, colonne] -> Forma canonica di selezione nel df
 
df.loc[df["Product_Name"] == "Cabbage", "Category"] = "Fruits & Vegetables"


In [11]:
# Controllo della correzione

df.loc[
    df["Product_Name"] == "Cabbage",
    ["Product_ID", "Product_Name", "Category"]
]


,Product_ID,Product_Name,Category
23,67-025-1245,Cabbage,Fruits & Vegetables
501,21-013-3508,Cabbage,Fruits & Vegetables
521,79-428-8753,Cabbage,Fruits & Vegetables
652,45-380-4627,Cabbage,Fruits & Vegetables
685,10-378-9729,Cabbage,Fruits & Vegetables
788,75-927-9108,Cabbage,Fruits & Vegetables
806,82-538-4809,Cabbage,Fruits & Vegetables
987,31-745-6850,Cabbage,Fruits & Vegetables


In [12]:
# Controllo delle righe duplicate

numero_duplicati = df.duplicated().sum()

print("Numero di righe duplicate:")
print(numero_duplicati)


Numero di righe duplicate:
0


In [13]:
# Controllo dei Product_ID duplicati

duplicati_product_id = df["Product_ID"].duplicated().sum()

print("Numero di Product_ID duplicati:")
print(duplicati_product_id)


Numero di Product_ID duplicati:
0


In [14]:

print("Valori mancanti in Category:")
print(df["Category"].isna().sum())


Valori mancanti in Category:
0


In [15]:
# Elenco  dei valori unici

print("Categorie:")
print(sorted(df["Category"].dropna().unique()))

print("\nStati:")
print(sorted(df["Status"].dropna().unique()))


# Controllo dei valori presenti nelle colonne categoriche

print("Valori presenti nella colonna Category:")
print(df["Category"].value_counts(dropna=False))

print("\nValori presenti nella colonna Status:")
print(df["Status"].value_counts(dropna=False))


Categorie:
['Bakery', 'Beverages', 'Dairy', 'Fruits & Vegetables', 'Grains & Pulses', 'Oils & Fats', 'Seafood']

Stati:
['Active', 'Backordered', 'Discontinued']
Valori presenti nella colonna Category:
Category
Fruits & Vegetables    332
Dairy                  180
Grains & Pulses        162
Seafood                 90
Oils & Fats             77
Beverages               75
Bakery                  74
Name: count, dtype: int64

Valori presenti nella colonna Status:
Status
Discontinued    333
Active          332
Backordered     325
Name: count, dtype: int64


In [16]:
# Pulizia di eventuali spazi iniziali o finali nelle colonne categoriche

colonne_testo = [
    "Product_ID",
    "Product_Name",
    "Category",
    "Supplier_ID",
    "Supplier_Name",
    "Warehouse_Location",
    "Status"
]

for colonna in colonne_testo:
    df[colonna] = df[colonna].str.strip()

# Verifica successiva alla pulizia

print("Valori presenti in Category:")
print(df["Category"].value_counts(dropna=False))

print("\nValori presenti in Status:")
print(df["Status"].value_counts(dropna=False))

Valori presenti in Category:
Category
Fruits & Vegetables    332
Dairy                  180
Grains & Pulses        162
Seafood                 90
Oils & Fats             77
Beverages               75
Bakery                  74
Name: count, dtype: int64

Valori presenti in Status:
Status
Discontinued    333
Active          332
Backordered     325
Name: count, dtype: int64


In [17]:
# Controllo iniziale della colonna Unit_Price

print("Tipo di dato di Unit_Price:")
print(df["Unit_Price"].dtype)

print("\nPrimi 5 valori:")
print(df["Unit_Price"].head(5))

print("\nValori mancanti:")
print(df["Unit_Price"].isna().sum())


Tipo di dato di Unit_Price:
str

Primi 5 valori:
0     $4.50 
1    $20.00 
2     $6.00 
3     $1.50 
4     $4.00 
Name: Unit_Price, dtype: str

Valori mancanti:
0


In [18]:
# Rimozione di $ e conversione della colonna Unit_Price in formato numerico

df["Unit_Price"] = pd.to_numeric(df["Unit_Price"].astype(str).str.replace("$", "", regex=False).str.strip(), errors="coerce")

# Verifica del tipo di dato e di eventuali valori mancanti

print("Tipo di dato:", df["Unit_Price"].dtype)
print("Conversioni non riuscite:", df["Unit_Price"].isna().sum())
print(df["Unit_Price"].head(10))


Tipo di dato: float64
Conversioni non riuscite: 0
0     4.50
1    20.00
2     6.00
3     1.50
4     4.00
5     1.75
6     2.50
7     2.50
8     3.00
9     1.00
Name: Unit_Price, dtype: float64


In [19]:
# Controllo iniziale delle colonne contenenti date

print("Tipo di dato di Date_Received:")
print(df["Date_Received"].dtype)

print("\nTipo di dato di Last_Order_Date:")
print(df["Last_Order_Date"].dtype)

print("\nTipo di dato di Expiration_Date:")
print(df["Expiration_Date"].dtype)

print("\nPrime 20 righe delle colonne data:")
print(df[["Date_Received", "Last_Order_Date", "Expiration_Date"]].head(20))

Tipo di dato di Date_Received:
str

Tipo di dato di Last_Order_Date:
str

Tipo di dato di Expiration_Date:
str

Prime 20 righe delle colonne data:
   Date_Received Last_Order_Date Expiration_Date
0      8/16/2024       6/29/2024       9/19/2024
1      11/1/2024       5/29/2024        5/8/2024
2       8/3/2024       6/10/2024       9/22/2024
3      12/8/2024       2/19/2025       4/17/2024
4       7/3/2024      10/11/2024       10/5/2024
5      12/3/2024       5/26/2024        9/5/2024
6      3/18/2024        5/7/2024       6/20/2024
7       2/3/2025        4/9/2024        2/5/2025
8      12/4/2024        6/2/2024        1/8/2025
9     11/18/2024      11/14/2024        7/8/2024
10    12/18/2024       5/24/2024        1/7/2025
11    10/19/2024      12/11/2024      10/14/2024
12    12/13/2024        2/9/2025       7/18/2024
13      4/5/2024       9/21/2024        5/6/2024
14     3/11/2024       6/30/2024       6/14/2024
15     2/27/2024        4/2/2024       10/5/2024
16    11/16/2024    

In [20]:
# Conversione delle colonne data dal formato testo MM/DD/YYYY al formato datetime

colonne_data = ["Date_Received", "Last_Order_Date", "Expiration_Date"]

for colonna in colonne_data:
    df[colonna] = pd.to_datetime(df[colonna], format="%m/%d/%Y", errors="coerce")
    

In [21]:
# Verifica del tipo di dato e di eventuali errori di conversione nelle colonne data

print(df[colonne_data].dtypes)

print("\nValori mancanti dopo la conversione:")
print(df[colonne_data].isna().sum())

print("\nPrime 20 righe:")
print(df[colonne_data].head(20))


Date_Received      datetime64[us]
Last_Order_Date    datetime64[us]
Expiration_Date    datetime64[us]
dtype: object

Valori mancanti dopo la conversione:
Date_Received      0
Last_Order_Date    0
Expiration_Date    0
dtype: int64

Prime 20 righe:
   Date_Received Last_Order_Date Expiration_Date
0     2024-08-16      2024-06-29      2024-09-19
1     2024-11-01      2024-05-29      2024-05-08
2     2024-08-03      2024-06-10      2024-09-22
3     2024-12-08      2025-02-19      2024-04-17
4     2024-07-03      2024-10-11      2024-10-05
5     2024-12-03      2024-05-26      2024-09-05
6     2024-03-18      2024-05-07      2024-06-20
7     2025-02-03      2024-04-09      2025-02-05
8     2024-12-04      2024-06-02      2025-01-08
9     2024-11-18      2024-11-14      2024-07-08
10    2024-12-18      2024-05-24      2025-01-07
11    2024-10-19      2024-12-11      2024-10-14
12    2024-12-13      2025-02-09      2024-07-18
13    2024-04-05      2024-09-21      2024-05-06
14    2024-03-11  

In [22]:
# Controllo delle possibili incoerenze tra ricezione e scadenza

# Righe in cui la scadenza è precedente alla ricezione

date_non_coerenti = df[
    df["Expiration_Date"] < df["Date_Received"]
]

numero_non_coerenti = len(date_non_coerenti)

#risultato

print("Totale righe:", len(df))
print("Possibili incoerenze tra ricezione e scadenza:", numero_non_coerenti)


Totale righe: 990
Possibili incoerenze tra ricezione e scadenza: 496


---
Sono state individuate 496 potenziali incoerenze temporali, nelle quali la data di scadenza precede la data dell’ultimo ricevimento. Poiché il dataset non contiene un identificativo del lotto, questi record vengono segnalati per verifica e non corretti.

In [23]:
# Controllo dei valori numerici non validi o poco plausibili

# Selezioniamo le righe con quantità di magazzino negative

stock_negativo = df[df["Stock_Quantity"] < 0]

# Selezioniamo le righe con livello di riordino negativo

riordino_negativo = df[df["Reorder_Level"] < 0]

# Selezioniamo le righe con quantità da riordinare negativa

quantita_riordino_negativa = df[df["Reorder_Quantity"] < 0]

# Selezioniamo le righe con prezzo nullo o negativo

prezzo_non_valido = df[df["Unit_Price"] <= 0]

# Risultati

print("Stock negativo:", len(stock_negativo))
print("Livello di riordino negativo:", len(riordino_negativo))
print("Quantità di riordino negativa:", len(quantita_riordino_negativa))
print("Prezzo nullo o negativo:", len(prezzo_non_valido))


Stock negativo: 0
Livello di riordino negativo: 0
Quantità di riordino negativa: 0
Prezzo nullo o negativo: 0


In [24]:
# Verifica finale del dataset prima dell'esportazione

print("DIMENSIONI DEL DATASET")
print("Righe:", df.shape[0])
print("Colonne:", df.shape[1])

print("\nTIPI DI DATO")
print(df.dtypes)

print("\nVALORI MANCANTI")
print(df.isna().sum())

print("\nRIGHE DUPLICATE")
print(df.duplicated().sum())

print("\nDUPLICATI NEL PRODUCT_ID")
print(df["Product_ID"].duplicated().sum())

DIMENSIONI DEL DATASET
Righe: 990
Colonne: 16

TIPI DI DATO
Product_ID                            str
Product_Name                          str
Category                              str
Supplier_ID                           str
Supplier_Name                         str
Stock_Quantity                      int64
Reorder_Level                       int64
Reorder_Quantity                    int64
Unit_Price                        float64
Date_Received              datetime64[us]
Last_Order_Date            datetime64[us]
Expiration_Date            datetime64[us]
Warehouse_Location                    str
Sales_Volume                        int64
Inventory_Turnover_Rate             int64
Status                                str
dtype: object

VALORI MANCANTI
Product_ID                 0
Product_Name               0
Category                   0
Supplier_ID                0
Supplier_Name              0
Stock_Quantity             0
Reorder_Level              0
Reorder_Quantity           0
Unit_

In [27]:
# Statistiche descrittive delle colonne numeriche del dataset

print(df.describe())

       Stock_Quantity  Reorder_Level  Reorder_Quantity  Unit_Price  \
count      990.000000     990.000000        990.000000  990.000000   
mean        55.609091      51.215152         51.913131    5.924192   
min         10.000000       1.000000          1.000000    0.200000   
25%         33.000000      25.250000         25.000000    2.500000   
50%         56.000000      53.000000         54.000000    4.225000   
75%         79.000000      77.000000         77.000000    7.000000   
max        100.000000     100.000000        100.000000   98.430000   
std         26.300775      29.095241         29.521059    6.491280   

                    Date_Received             Last_Order_Date  \
count                         990                         990   
mean   2024-08-23 02:18:10.909091  2024-08-25 19:20:43.636363   
min           2024-02-25 00:00:00         2024-02-25 00:00:00   
25%           2024-05-27 00:00:00         2024-05-29 00:00:00   
50%           2024-08-19 00:00:00         20

---
## Prime osservazioni sulle variabili numeriche

Le principali variabili quantitative mostrano distribuzioni abbastanza regolari, con valori medi e mediani generalmente vicini.

Le quantità di stock, i livelli di riordino, le quantità di riordino, il volume delle vendite e il tasso di rotazione presentano intervalli ampi. Dalla sola tabella descrittiva non emergono valori impossibili, ma le distribuzioni verranno approfondite con analisi specifiche.

La variabile `Unit_Price` mostra una maggiore asimmetria: il prezzo medio è superiore alla mediana e il valore massimo è molto distante dal terzo quartile. Questo suggerisce la presenza di alcuni prodotti con prezzo unitario molto più elevato rispetto alla maggior parte del catalogo.

Le colonne temporali coprono un intervallo di circa un anno, da febbraio 2024 a febbraio 2025.

In [29]:
# Calcolo dei limiti per individuare i possibili outlier di Unit_Price

Q1 = df["Unit_Price"].quantile(0.25)
Q3 = df["Unit_Price"].quantile(0.75)

IQR = Q3 - Q1

limite_inferiore = Q1 - 1.5 * IQR
limite_superiore = Q3 + 1.5 * IQR

print("Primo quartile:", round(Q1, 2))
print("Terzo quartile:", round(Q3, 2))
print("Intervallo interquartile:", round(IQR, 2))
print("Limite inferiore:", round(limite_inferiore, 2))
print("Limite superiore:", round(limite_superiore, 2))

Primo quartile: 2.5
Terzo quartile: 7.0
Intervallo interquartile: 4.5
Limite inferiore: -4.25
Limite superiore: 13.75


In [30]:
# Selezione dei prodotti con prezzo superiore alla soglia statistica

outlier_prezzo = df[
    df["Unit_Price"] > limite_superiore
].copy()

print("Numero di prodotti con prezzo estremo:", len(outlier_prezzo))

Numero di prodotti con prezzo estremo: 75


In [31]:
# Visualizzazione dei prodotti con prezzo superiore alla soglia statistica

display(
    outlier_prezzo[
        [
            "Product_ID",
            "Product_Name",
            "Category",
            "Unit_Price"
        ]
    ]
    .sort_values(
        by="Unit_Price",
        ascending=False
    )
    .head(20)
)

,Product_ID,Product_Name,Category,Unit_Price
113,87-698-0944,Banana,Fruits & Vegetables,98.43
685,10-378-9729,Cabbage,Fruits & Vegetables,66.55
325,48-957-8596,Banana,Fruits & Vegetables,53.82
208,68-734-1585,Greek Yogurt,Fruits & Vegetables,51.17
699,32-270-1385,Greek Yogurt,Fruits & Vegetables,42.58
161,80-374-5711,Herbal Tea,Beverages,31.00
526,39-192-3521,Herbal Tea,Beverages,30.50
630,19-244-9890,Herbal Tea,Beverages,30.00
235,27-442-3654,Herbal Tea,Beverages,30.00
33,79-136-9840,Herbal Tea,Beverages,30.00


### Interpretazione

Il metodo IQR individua come possibili valori estremi i prezzi superiori al limite calcolato.

Questi valori non vengono eliminati, perché possono rappresentare prodotti realmente più costosi. Vengono mantenuti e segnalati per le analisi successive.

## Analisi delle variabili categoriche

In questa sezione vengono analizzate le principali variabili qualitative del dataset, con particolare attenzione alla categoria del prodotto, allo stato e alla posizione di magazzino.

In [33]:
# Individuazione delle colonne categoriche e conteggio dei valori distinti

colonne_categoriche = df.select_dtypes(
    include=["object", "string"]
).columns

print("Colonne categoriche presenti nel dataset:\n")

for colonna in colonne_categoriche:
    print(
        colonna,
        "- valori distinti:",
        df[colonna].nunique()
    )

Colonne categoriche presenti nel dataset:

Product_ID - valori distinti: 990
Product_Name - valori distinti: 121
Category - valori distinti: 7
Supplier_ID - valori distinti: 990
Supplier_Name - valori distinti: 350
Warehouse_Location - valori distinti: 990
Status - valori distinti: 3


In [25]:
# Esportazione del dataset pulito in un nuovo file CSV

file_output = Path("Dataset_pulito.csv")

df.to_csv(
    file_output,
    index=False,
    date_format="%Y-%m-%d"
)

print("Dataset esportato correttamente.")
print("Percorso del file:", file_output.resolve())

Dataset esportato correttamente.
Percorso del file: C:\Users\mace\OneDrive\python\installazione1\project work\Analisi dati di magazzino\Dataset_pulito.csv


In [26]:
# Verifica dell'esistenza del file esportato

print("Il file esiste:", file_output.exists())

Il file esiste: True
